src/components/structures/TimelinePanel.tsx

```diff
@@ -369,9 +396,55 @@ class TimelinePanel extends React.Component<IProps, IState> {
             client.removeListener(MatrixEventEvent.Replaced, this.onEventReplaced);
             client.removeListener(MatrixEventEvent.VisibilityChange, this.onEventVisibilityChange);
             client.removeListener(ClientEvent.Sync, this.onSync);
+            client.removeListener(ClientEvent.Sync, this.onDumpDebugLogs);
         }
     }
 
+    private onDumpDebugLogs = (): void => {
+        const roomId = this.props.timelineSet.room.roomId;
+        // This includes state and hidden events which we don't render
+        const eventIdList = this.state.events.map((ev) => ev.getId());
+
+        // Get the list of actually rendered events seen in the DOM.
+        let renderedEventIds;
+        const messagePanel = this.messagePanel.current;
+        if (messagePanel) {
+            const messagePanelNode = ReactDOM.findDOMNode(messagePanel) as HTMLElement;
+            if (messagePanelNode) {
+                const actuallyRenderedEvents = messagePanelNode.querySelectorAll('[data-event-id]');
+                renderedEventIds = [...actuallyRenderedEvents].map((renderedEvent) => {
+                    return renderedEvent.getAttribute('data-txn-id') || renderedEvent.getAttribute('data-event-id');
+                });
+            }
+        }
+
+        let serializedEventIdsFromTimelineSets;
+        let serializedEventIdsFromThreadsTimelineSets;
+        let serializedThreadsMap = {};
+        const client = MatrixClientPeg.get();
+        if (client) {
+            const room = client.getRoom(roomId);
+            const timelineSets = room.getTimelineSets();
+            const threadsTimelineSets = room.threadsTimelineSets;
+
+            serializedEventIdsFromTimelineSets = serializeEventIdsFromTimelineSets(timelineSets);
+            serializedEventIdsFromThreadsTimelineSets = serializeEventIdsFromTimelineSets(threadsTimelineSets);
+
+            room.getThreadsMap().forEach((thread, threadId) => {
+                serializedThreadsMap[threadId] = thread.events.map(ev => ev.getId());
+            });
+        }
+
+        logger.debug(
+            `TimelinePanel(${this.context.timelineRenderingType}): Debugging info for ${roomId}\n` +
+            `\tevents(${eventIdList.length})=${JSON.stringify(eventIdList)}\n` +
+            `\trenderedEventIds(${renderedEventIds ? renderedEventIds.length : 0})=${JSON.stringify(renderedEventIds)}\n` +
+            `\tserializedEventIdsFromTimelineSets=${JSON.stringify(serializedEventIdsFromTimelineSets)}\n` +
+            `\tserializedEventIdsFromThreadsTimelineSets=${JSON.stringify(serializedEventIdsFromThreadsTimelineSets)}\n` +
```

Comment:
Confusingly, `room.threadsTimelineSets`-> `serializedEventIdsFromThreadsTimelineSets` are only created and used for the `ThreadPanel` where threads are listed not an individual thread.

Thought it was relevant to log but could also be moved to `ThreadPanel` instead.

Reply:
Re-commented on the diff since this one got outdated, https://github.com/matrix-org/matrix-react-sdk/pull/8502#discussion_r866366356